### PR TITLE
Introduce async closing of `IndexShard`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Closing an `IndexShard` may be expensive if the `IndexWriter` takes a
long time to roll back and shut down. This commit adds the facility for
forking the expensive work onto a different thread, although in
production it still runs this work directly.

Relates ES-8334

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Extracted from https://github.com/elastic/elasticsearch/pull/107952 the API changes and some supporting assertions, although ultimately this PR still doesn't fork the work off the applier thread yet. But I'm pretty sure this direction is going to get us there.

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 1]
```diff
@@ -583,17 +619,34 @@ private void closeShard(String reason, ShardId sId, IndexShard indexShard, Store
...
-                        final boolean flushEngine = deleted.get() == false && closed.get();
-                        indexShard.close(reason, flushEngine);
-                    } catch (Exception e) {
-                        logger.debug(() -> "[" + shardId + "] failed to close index shard", e);
-                        // ignore
-                    }
+                    // only flush if we are closed (closed index or shutdown) and if we are not deleted
+                    final boolean flushEngine = deleted.get() == false && closed.get();
+                    // if the store is still open, want to keep it open while closing the shard
+                    final var hasStoreRef = store != null && store.tryIncRef();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/IndexService.java:625]: Do we need the incref/decref of store?

I would think that if the shard has an engine, it holds the necessary ref. If it does not, no ref is needed.

or in other words, the `IndexShard` instance should be reponsible for the store refs it needs itself?

**Review by henningandersen — COMMENTED** [review]
Looks good. Did not find any obvious extra validation points. I'll see if something springs to mind, otherwise, I'll lgtm it on the next round.

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 2]
```diff
@@ -583,17 +619,34 @@ private void closeShard(String reason, ShardId sId, IndexShard indexShard, Store
...
-                        // only flush if we are closed (closed index or shutdown) and if we are not deleted
-                        final boolean flushEngine = deleted.get() == false && closed.get();
-                        indexShard.close(reason, flushEngine);
-                    } catch (Exception e) {
-                        logger.debug(() -> "[" + shardId + "] failed to close index shard", e);
-                        // ignore
-                    }
+                    // only flush if we are closed (closed index or shutdown) and if we are not deleted
+                    final boolean flushEngine = deleted.get() == false && closed.get();
+                    // if the store is still open, want to keep it open while closing the shard
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/IndexService.java]: I'm being cautious. I think if the store exists here it should still be open indeed. Added an assertion to that effect, but would rather keep the defensive production code here.

File: server/src/main/java/org/elasticsearch/index/IndexService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/IndexService.java]: That's true, but previously we held the store open until after calling `afterIndexShardClosed` (and there was even a comment about doing so). I don't think it's important to do this looking at the implementations but, again, trying not to change semantics unnecessarily here.

File: server/src/main/java/org/elasticsearch/index/CloseUtils.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+public class CloseUtils {
+
+    private CloseUtils() {/* no instances */}
+
+    /**
+     * Execute a naturally-async action (e.g. to close a shard) but using the current thread so that it completes synchronously, re-throwing
+     * any exception that might be passed to its listener.
+     */
+    public static void executeDirectly(CheckedConsumer<ActionListener<Void>, IOException> action) throws IOException {
+        final var closeExceptionRef = new AtomicReference<Exception>();
```
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/CloseUtils.java:31]: I think the `assertOnce` fix is neat indeed to validate that the action listener is invoked. What I am after here however is that the listener has been invoked before this method exits, i.e., that we do not somewhere in the future spawn this inadvertently without noticing.

Can we either add a sentinel exception object to set in `onResponse` or an AtomicBoolean - just for asserting on it?

**Review by henningandersen — APPROVED** [review]
LGTM.

I left one stray comment (outside this review), feel free to ignore if you prefer.

I could not come up with good extra validations for this part, it essentially is more of a refactor/api change than a change of how things work. We could consider to in the future add something to ensure we only do lucene directory operations on the applier thread when we expect to, but that seems separate to this.

File: server/src/main/java/org/elasticsearch/index/CloseUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/CloseUtils.java:42]: :+1: yeah sorry I missed that part of your comment. Added in 3ccdc7dee0e.

[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/CloseUtils.java:42]: :thu: thanks

**Comment by StephanErb** [comment]
Are we only expecting a benefit for active write shards, or could this also bring an improvement to slow cluster state updates on cold nodes?

**Comment by DaveCTurner** [comment]
Hi @StephanErb, thanks for your question. This PR doesn't change any behaviour or performance. I'm happy to discuss broader topics in this area, but would prefer to keep PR comments related to the PR itself. If you want to know more, would you open a thread in the [Elasticsearch forum](https://discuss.elastic.co/c/elasticsearch)? Link this PR so other readers have useful context, and feel free to mention me (I'm `@DavidTurner`) there too to make sure I don't miss it.

**Pull request merged**